Outdated: GRPC connect + /projects

thomas-mauran opened the pull request:
## GRPC connect + /projects

adding the following routes:
[GET] projects/ -> return all of the projects
[GET] projects/uuid -> info of project

[POST] projects/ -> creates a project

[DELETE] projects/uuid -> delete a project (both db and grpc)

### Additional changes

* Modifying db projects user_id to not be unique
* adding info in the jwt for connection
* adding custom decorator for jwt info recuperation

Commit:
feat(diablox9): modern warfare 2

Commit:
feat(diablox9): modern warfare 2

Commit:
feat(diablox9): modern warfare 2

mfernd commented on src/projects/projects.service.ts:
```diff
@@ -14,33 +14,84 @@ export class ProjectsService {
+    // If the user is admin we return every project in the database
+    if (user.isAdmin) {
+      const projects = await this.prisma.project.findMany();
+      return projects.map((project) => project);
```
`sanitizeOutput()` or not `map(), that is the question

(I prefer with sanitize, to have the same data structure on the front)

thomas-mauran replied:
``` return projects.map((project) => this.sanitizeOutput(project));```
done :+1:

mfernd commented on src/projects/projects.controller.ts:
```diff
@@ -1,48 +1,78 @@
+      throw new BadRequestException('Project already exists');
     }
 
-    // DELETE /projects/:uuid
-    // This action deletes a #${id} project
-    @Delete(':uuid')
-    async delete(@Param('uuid') uuid: string) {
-        this.projectsService.delete(uuid);
-        return `OK`;
-    }
+    const createdProject = await this.projectsService.create(user.id, request);
+    // Create the project's repository
+    const projectPath = `${createdProject.uuid}`;
+    const repositoryRequest: RepositoryRequest = this.createRepositoryRequest(projectPath);
+
+    // Send the gRPC request to create the repository
+    await this.gitRepoManagerService.create(repositoryRequest);
+
+    return createdProject;
+  }
```
Can you do the prisma query inside a transaction (example: do not create a project in db if we fail to create the git repo)?

The same could be done for `delete`

Resource: [Prisma transaction api](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions)

mfernd commented on src/projects/projects.service.ts:
```diff
@@ -14,33 +14,84 @@ export class ProjectsService {
-    const project = await this.prisma.project.findUnique({
-      // FIXME: ensure one can only find their projects and not somebody else's
-      where: {uuid},
+  async findOne(uuid: string, userId: number): Promise<SanitizedProject | null> {
```
Could you throw an error if the project is not found instead of returning `null`?

Same in other methods

Resource: [findFirstOrThrow](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#findfirstorthrow)

thomas-mauran replied:
nice ! i replaced all the logic for findFirst to findFirstOrThrow

mfernd replied:
don't forget what you return from the function

noetarbouriech commented on src/pomegranate/pomegranate.controller.ts:
Don't leave everything here as comment

thomas-mauran replied:
should i delete the file for this pr ?

mfernd replied:
yeah if you think it's better to do it in another PR, remove the module

BioTheWolff commented:
PR marked as draft for the moment, as there are files fully commented out, and files quoted "not working yet". I will gladly re-open the PR as soon as the code is ready for review!

thomas-mauran commented:
> PR marked as draft for the moment, as there are files fully commented out, and files quoted "not working yet". I will gladly re-open the PR as soon as the code is ready for review!

I deleted the file for now since it shouldn't be in this pr

Commit:
chore(pomegrenade): removing the unused file

Commit:
fix(projects): sanitize output findAll

sea-gull-diana commented on src/projects/projects.controller.ts:
```diff
@@ -1,48 +1,78 @@
+  // DELETE /projects/:uuid
+  // This action deletes a #${id} project
+  @Delete(':uuid')
+  @UseGuards(AuthGuard)
```
It shouldn't normally be necessary to protect routes with `@UseGuards`, they are all protected by default. If you want to have an unprotected route, add `@Public()` decorator.

thomas-mauran replied:
done :+1:

BioTheWolff commented on src/pomegranate/pomegranate.controller.ts:
Please remove file, as it is full commented and thus not needed

thomas-mauran replied:
done :+1:

sea-gull-diana commented on src/projects/projects.controller.ts:
```diff
@@ -1,48 +1,78 @@
+  @UseGuards(AuthGuard)
+  async delete(@Query() request: DeleteRepositoryDto, @GetUser() user: UserDecoratorType): Promise<SanitizedProject> {
+    // Delete the project from the database
+    const project = await this.projectsService.delete(request.uuid);
```
It is also necessary to verify that the project that is being read, modified or deleted belongs to the user who sent the request.

thomas-mauran replied:
Done :+1:

BioTheWolff commented on src/auth/interfaces/accessToken.ts:
```diff
@@ -1,3 +1,8 @@
+  accessToken: string;
+}
+
+export interface UserLogin {
```
What is this interface doing?

thomas-mauran replied:
if is used in the login service as asked by some frontend devs to send back not only the token but also the isAdmin 
```js
async login(credentials: LoginUserDto): Promise<UserLogin> {
    const user = await this.usersService.findOneUnsanitized({ email: credentials.email });
    const passwordHash = user.password;
    const isCorrectPassword = await bcrypt.compare(credentials.password, passwordHash);
    if (!isCorrectPassword || !passwordHash) {
      throw new UnauthorizedException();
    }

    // Add a username in the JWT token
    const payload = { username: user.username, isAdmin: user.isAdmin, id: user.id };

    return { accessToken: await this.jwtService.signAsync(payload), isAdmin: user.isAdmin };

BioTheWolff commented on src/projects/projects.controller.ts:
```diff
@@ -1,48 +1,78 @@
-        // TODO: implement and remove the 501 decorator
-        return '501 Not Implemented';
+  @Post()
+  @UsePipes(new ValidationPipe())
```
`ValidationPipe` is set as global, see `main.ts` file for confirmation

thomas-mauran replied:
nice, done :+1:

BioTheWolff commented on src/projects/projects.controller.ts:
```diff
@@ -1,48 +1,78 @@
+  // DELETE /projects/:uuid
+  // This action deletes a #${id} project
+  @Delete(':uuid')
+  @UseGuards(AuthGuard)
```
Guard is set by default and thus not needed

thomas-mauran replied:
done :+1:

Commit:
fix(projects): use findFirstOrThrow instead of findFirst

sea-gull-diana reviewed (commented):
Nice :+1:  I'm not sure why your are calling a controller for managing deployments "pomegranate", why not just `deployment.controller`? I suggest creating a whole module for deployments and moving `pomegranate.service` there from the "projects" folder :)

mfernd committed:
fix(errors): use throw errors instead of null...

mfernd committed:
chore: remove unused imports

mfernd committed:
chore: remove unused imports

mfernd committed:
fix: remove authguard + add paramater `user.id`

mfernd committed:
fix(projects): simplify catch indents

Commit:
fix(projects): fix find ... nestjs moment

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

Commit:
refactor(projects): use new id field

Commit:
fix(projects): use transaction thx matéo

Commit:
fix(projects): use transaction thx matéo

Commit:
chore(projects): remove  unused pomegrenade service

sea-gull-diana committed:
feat: use passport-jwt for authentication (#9)

Modifies user authentication by using passport-jwt strategy.

Add refresh tokens and possibly API tokens (for CLI login).

---------

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

sea-gull-diana committed:
OpenAPI spec and Swagger UI (#13)

Initializes Swagger using the SwaggerModule, allows using Swagger UI for
testing and autogenerates OpenAPI spec.

- run `npm i`
- delete *dist* folder
- run `npm run start:dev`

See if we can figure out a way to make the OpenAPI spec more precise
regarding the endpoints return types and codes.

Sturmwoelfin committed:
feat: added mail module

Added Functionalities

- email verification
- password reset

---------
Co-authored-by: Muriel Paraire <muriel.paraire@etu.umontpellier.fr>
Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>
Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

noetarbouriech committed:
feat: add basic Dockerfile for the api (#14)

This PR adds a basic `Dockerfile` with a build step and a run step.

---------

Co-authored-by: Lapsus Ordonné <22498591+lapsus-ord@users.noreply.github.com>

Commit:
feat(diablox9): modern warfare 2

Commit:
refactor(projects): use new id field

Commit:
chore: end my suffering
